onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -1407,6 +1407,369 @@ with the exception that numpy default keepdims to False instead of True.)DOC")
           "Keep the reduced dimension or not, default 1 mean keep reduced dimension.",
           AttributeProto::INT);
 
+  ONNX_CONTRIB_OPERATOR_SCHEMA(QLinearAdd)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetDoc(R"DOC(Performs element-wise binary quantized addition (with Numpy-style broadcasting support).
+This operator supports **multidirectional (i.e., Numpy-style) broadcasting**.
+```
+C = (A_scale * (A - A_zero_point) + B_scale * (B - B_zero_point))/C_scale + C_zero_point
+```
+)DOC")
+      .Input(0, "A", "First operand.", "T")
+      .Input(
+          1,
+          "A_scale",
+          "Input A's scale. It's a scalar, which means a per-tensor/layer quantization.",
+          "tensor(float)")
+      .Input(
+          2,
+          "A_zero_point",
+          "Input A zero point. Default value is 0 if it's not specified. It's a scalar, which means a per-tensor/layer quantization.",
+          "T",
+          OpSchema::Optional)
+      .Input(3, "B", "Second operand.", "T")
+      .Input(
+          4,
+          "B_scale",
+          "Input B's scale. It's a scalar, which means a per-tensor/layer quantization.",
+          "tensor(float)")
+      .Input(
+          5,
+          "B_zero_point",
+          "Input B zero point. Default value is 0 if it's not specified. It's a scalar, which means a per-tensor/layer quantization.",
+          "T",
+          OpSchema::Optional)
+      .Input(
+          6,
+          "C_scale",
+          "Output scale. It's a scalar, which means a per-tensor/layer quantization.",
+          "tensor(float)")
+      .Input(
+          7,
+          "C_zero_point",
+          "Output zero point. Default value is 0 if it's not specified. It's a scalar, which means a per-tensor/layer quantization.",
+          "T",
+          OpSchema::Optional)
+      .Output(0, "C", "Result, has same element type as two inputs", "T")
+      .TypeConstraint("T", {"tensor(uint8)", "tensor(int8)"}, "Constrain input and output types to 8 bit signed and unsigned tensors.");
+
+  ONNX_CONTRIB_OPERATOR_SCHEMA(QLinearReduceMean)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetDoc(R"DOC(
+Computes the mean of the low-precision input tensor's element along the provided axes.
+The resulting tensor has the same rank as the input if keepdims equal 1. If keepdims equal 0,
+then the resulting tensor have the reduced dimension pruned. The above behavior is similar to numpy,
+with the exception that numpy default keepdims to False instead of True.
+Input and Output scales and zero points are used to requantize the output in a new range. 
+This helps to improve accuracy as after ReduceMean operation the range of the output is expected to decrease.
+
+```
+reduced_intermediate  = onnx.ReduceMean(data, axis=axes, keepdims=keepdims == 1)
```

Comment:
Yes it can be reduced in it's quantized state... infact I am going to add 8 bit support to the existing ReduceMean op in onnx... the difference between the ONNX op and this new op will be that with this op the range of output can be changed post reduction. 

During reduction the intermediate result of the operation will be held in an int32 accumulator....

Reply:
That is true but if you perform the divide by number of reduction elements, in int space, then the result will be different because you lose additional precision from the int8 intermediate. If you reduce from int8->int8 output with int32 accumulators you will not get the same result as if you dequantized then reduced in float then quantized.

Reply:
You could also get different results if you reduced int8->int32 as well. It would however be less obvious.
The math required to avoid doing a reduction in float space requires that the division of the number of elements be done in float space when the scale factors are accounted for.